File: ships-backend/app/utils/debounced_logger.py
"""
Debounced logging utility to prevent log spam.
Tracks recent log messages and only emits if sufficient time has passed.
"""
import time
import logging
from typing import Dict, Tuple
from functools import wraps

class DebouncedLogger:
    """Logger wrapper that debounces repeated messages."""
    
    def __init__(self, logger: logging.Logger, debounce_seconds: float = 1.0):
        self.logger = logger
        self.debounce_seconds = debounce_seconds
        self._last_log_times: Dict[Tuple[str, str], float] = {}
        self._log_counts: Dict[Tuple[str, str], int] = {}
    
    def _should_log(self, level: str, message: str) -> bool:
        """Check if enough time has passed since last identical log."""
        key = (level, message)
        now = time.time()
        
        last_time = self._last_log_times.get(key, 0)
        time_since_last = now - last_time
        
        if time_since_last >= self.debounce_seconds:
            # Log the count if this message was suppressed
            count = self._log_counts.get(key, 0)
            if count > 0:
                self._log_counts[key] = 0
                return True  # Will add count suffix
            
            self._last_log_times[key] = now
            return True
        
        # Increment suppressed count
        self._log_counts[key] = self._log_counts.get(key, 0) + 1
        return False
    
    def _get_log_func(self, level: str):
        """Get the appropriate logging function."""
        return getattr(self.logger, level.lower())
    
    def log(self, level: str, message: str, *args, **kwargs):
        """Log a message with debouncing."""
        key = (level, message)
        
        if self._should_log(level, message):
            count = self._log_counts.get(key, 0)
            if count > 0:
                message = f"{message} (suppressed {count} similar logs)"
                self._log_counts[key] = 0
            
            log_func = self._get_log_func(level)
            log_func(message, *args, **kwargs)
    
    def info(self, message: str, *args, **kwargs):
        self.log("INFO", message, *args, **kwargs)
    
    def debug(self, message: str, *args, **kwargs):
        self.log("DEBUG", message, *args, **kwargs)
    
    def warning(self, message: str, *args, **kwargs):
        self.log("WARNING", message, *args, **kwargs)
    
    def error(self, message: str, *args, **kwargs):
        self.log("ERROR", message, *args, **kwargs)


def debounced_log(debounce_seconds: float = 1.0):
    """
    Decorator to debounce logging within a function.
    
    Usage:
        @debounced_log(2.0)
        def my_function():
            logger.info("This won't spam")
    """
    def decorator(func):
        debouncer = None
        
        @wraps(func)
        def wrapper(*args, **kwargs):
            nonlocal debouncer
            
            # Intercept logger calls within the function
            # This is a simplified version - you'd typically patch the logger
            return func(*args, **kwargs)
        
        return wrapper
    return decorator

File: prism-frontend/prism-ai-web/src/utils/debouncedLogger.ts
/**
 * Debounced logging utility to prevent console spam.
 * Tracks recent log messages and only emits if sufficient time has passed.
 */

interface LogEntry {
  lastTime: number;
  count: number;
}

class DebouncedLogger {
  private debounceMs: number;
  private logCache: Map<string, LogEntry>;
  private flushInterval: number | null;

  constructor(debounceMs: number = 1000) {
    this.debounceMs = debounceMs;
    this.logCache = new Map();
    this.flushInterval = null;
    
    // Periodically flush suppressed logs
    this.startFlushInterval();
  }

  private startFlushInterval() {
    this.flushInterval = window.setInterval(() => {
      this.flushSuppressed();
    }, this.debounceMs * 2);
  }

  private flushSuppressed() {
    const now = Date.now();
    
    this.logCache.forEach((entry, key) => {
      const timeSinceLastLog = now - entry.lastTime;
      
      if (entry.count > 0 && timeSinceLastLog >= this.debounceMs) {
        console.log(`${key} (×${entry.count} suppressed)`);
        entry.count = 0;
      }
    });
  }

  private shouldLog(message: string): { should: boolean; count: number } {
    const now = Date.now();
    const cached = this.logCache.get(message);

    if (!cached) {
      this.logCache.set(message, { lastTime: now, count: 0 });
      return { should: true, count: 0 };
    }

    const timeSinceLastLog = now - cached.lastTime;

    if (timeSinceLastLog >= this.debounceMs) {
      const suppressedCount = cached.count;
      cached.lastTime = now;
      cached.count = 0;
      return { should: true, count: suppressedCount };
    }

    cached.count++;
    return { should: false, count: cached.count };
  }

  log(message: string, ...args: any[]) {
    const { should, count } = this.shouldLog(message);
    
    if (should) {
      if (count > 0) {
        console.log(`${message} (×${count} suppressed)`, ...args);
      } else {
        console.log(message, ...args);
      }
    }
  }

  info(message: string, ...args: any[]) {
    const { should, count } = this.shouldLog(message);
    
    if (should) {
      if (count > 0) {
        console.info(`${message} (×${count} suppressed)`, ...args);
      } else {
        console.info(message, ...args);
      }
    }
  }

  warn(message: string, ...args: any[]) {
    const { should, count } = this.shouldLog(message);
    
    if (should) {
      if (count > 0) {
        console.warn(`${message} (×${count} suppressed)`, ...args);
      } else {
        console.warn(message, ...args);
      }
    }
  }

  error(message: string, ...args: any[]) {
    // Never debounce errors - always show them
    console.error(message, ...args);
  }

  debug(message: string, ...args: any[]) {
    const { should, count } = this.shouldLog(message);
    
    if (should) {
      if (count > 0) {
        console.debug(`${message} (×${count} suppressed)`, ...args);
      } else {
        console.debug(message, ...args);
      }
    }
  }

  destroy() {
    if (this.flushInterval) {
      window.clearInterval(this.flushInterval);
      this.flushInterval = null;
    }
    this.flushSuppressed();
  }
}

// Export singleton instance
export const debouncedLogger = new DebouncedLogger(1000);

// Export class for custom instances
export { DebouncedLogger };
